### PR TITLE
Fixed map editor keeping removed owners on actors

### LINKED ISSUE #16868: Replace or remove actors of non-existing players when saving in the editor
As seen in #16867 you can easily place a new `mpspawn` and a few actors assigned to that player, then remove the spawn and the map will crash when loading the editor. Just remapping the owner of such units to an existing player (e.g. "Neutral" if it exists) on map save (to avoid losing actors when moving a spawn) sounds like a good idea to me. We can also consider removing such actors. (Perhaps this can be done if there are no valid players defined.) Opinions?

### PR DESCRIPTION
**Pull request by Mailaender**
Closes https://github.com/OpenRA/OpenRA/issues/16868

### CODE REVIEW

**Review by abcdefg30 — DISMISSED** [review]

**Comment by PunkPun** [comment]
I haven't tested the PR, but doesn't it feel wrong that actors are not saved? Instead something should happen on the removal of the owner, not at the moment of saving

**Comment by abcdefg30** [comment]
When the player is removed the owner of the actors gets reset to `Neutral` (right away, not when saving).

**Comment by Mailaender** [comment]
Changed the title to be less misleading.

**Comment by PunkPun** [comment]
> When the player is removed the owner of the actors gets reset to `Neutral` (right away, not when saving).

Actually thinking more about it. As we still don't have a tool for moving around actors (without deleting and adding) or mass setting an owner to loads of actors, this is definitely the wrong approach. If someone wants to move a spawn f.e. by one cell to the left he will now be forced to manually reset the owner to all actors. 

We should either wait for a move tool to be implemented or choose a different fix

**Comment by Mailaender** [comment]
This fixes a crash when playing the map, so I think this fix is justified. Removing the actors would be even more invasive. This is pretty unrelated to actor movement, although that would be a nice feature. Changing spawns is only relevant for multiplayer maps which usually come without pre-placed actors for the players anyway. If they do, then it is easily fixed without leaving the GUI.

**Comment by PunkPun** [comment]
> This fixes a crash when playing the map, so I think this fix is justified. 

I though it only crashed in the editor. When I tried it ingame it didn't crash

> Removing the actors would be even more invasive. 

There are many other methods I can think of to fix it. One would be fixing the editor UI to not crash on startup, or another solution would be to neutralise invalid owners only on saving. The best solution is of course adding a move tool + the current PR iteration

> Changing spawns is only relevant for multiplayer maps which usually come without pre-placed actors for the players anyway. If they do, then it is easily fixed without leaving the GUI.

It would not be easy if something like a 100 actors had custom owners, you'd have to go fixing 1 by 1 as we have no bulk functions. And this whole issue is about multiplayer maps no? How else are you gonna trigger the crash if you don't have multiple spawns

**Comment by Mailaender** [comment]
> I though it only crashed in the editor. When I tried it ingame it didn't crash

https://github.com/OpenHV/OpenHV/issues/647 describes the crash that follows.

> The best solution is of course adding a move tool + the current PR iteration

The main problem is my initial design #4496, where I tried to avoid settings and automatically guess the player number. An alternative would be to set the player number when creating a new map and not allowing a save until there are enough spawn points. However, that would be less convenient and complicated to redo, so I opted for a simple change. I don't see a reason to derail or postpone this.

**Comment by abcdefg30** [comment]
While this is not ideal, implementing a feature to manage players in the editor is Imo out of scope here. Most of the time people setting custom owners is unintentional or misinformed (that is why we have traits like [RequiresSpecificOwners](https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Traits/RequiresSpecificOwners.cs)), and most of the cases where players are "validly" set as owners are mission maps or similar which define their own players and often have no spawn points.

**Comment by PunkPun** [comment]
About about this solution https://github.com/OpenRA/OpenRA/compare/bleed...PunkPun:OpenRA:fix-editor ?

Only validating owners on save, so as to allow the user to move spawnpoints around without breaking everything

**Comment by Mailaender** [comment]
Hiding this on save will only provoke irritated people to submit another bug report for that behavior.

**Comment by PunkPun** [comment]
Which is fine. This isn't supposed to be a permanent solution. Only something we have till we introduce a move tool

**Comment by PunkPun** [comment]
https://github.com/OpenRA/OpenRA/compare/bleed...PunkPun:OpenRA:fix-editor2

Interestingly if I change the owner pre-serialisation it doesn't serialise correctly

**Comment by pchote** [comment]
IMO the cleaner solution here, which is what we already do with unknown terrain tiles (#18927), is to replace invalid owners on map *load*. We already have a lint test that reports invalid owners, so we don't lose anything by saving the invalid state.

**Comment by PunkPun** [comment]
Do you have this in mind? https://github.com/OpenRA/OpenRA/compare/bleed...PunkPun:OpenRA:fix-editor3

**Comment by pchote** [comment]
Depending on exactly what is crashing, a simpler and more general fix might be:
```diff
diff --git a/OpenRA.Game/Map/ActorInitializer.cs b/OpenRA.Game/Map/ActorInitializer.cs
index 2d700ec03a..d4a538d988 100644
--- a/OpenRA.Game/Map/ActorInitializer.cs
+++ b/OpenRA.Game/Map/ActorInitializer.cs
@@ -241,7 +241,7 @@ public OwnerInit(string value)
 
                public Player Value(World world)
                {
-                       return value ?? world.Players.First(x => x.InternalName == InternalName);
+                       return value ?? world.Players.FirstOrDefault(x => x.InternalName == InternalName) ?? world.WorldActor.Owner;
                }
 
                public void Initialize(MiniYaml yaml)
```

(but should also have a comment explaining the fallback!)

**Comment by PunkPun** [comment]
This does not fix the issue. The reflection present in this part of the codebase apparently prevents a proper stack trace

**Comment by pchote** [comment]
I see. It turns out that `Add` is using the `InternalName` directly, so the equivalent would be something like
```diff
diff --git a/OpenRA.Mods.Common/Traits/World/EditorActorLayer.cs b/OpenRA.Mods.Common/Traits/World/EditorActorLayer.cs
index ed92b2646d..c31f79f818 100644
--- a/OpenRA.Mods.Common/Traits/World/EditorActorLayer.cs
+++ b/OpenRA.Mods.Common/Traits/World/EditorActorLayer.cs
@@ -126,7 +126,10 @@ public IEnumerable<IRenderable> RenderAnnotations(Actor self, WorldRenderer wr)
 
                public EditorActorPreview Add(string id, ActorReference reference, bool initialSetup = false)
                {
-                       var owner = Players.Players[reference.Get<OwnerInit>().InternalName];
+                       var ownerName = reference.Get<OwnerInit>().InternalName;
+                       if (!Players.Players.TryGetValue(ownerName, out var owner))
+                               owner = Players.Players.Values.First(p => p.OwnsWorld);
+
                        var preview = new EditorActorPreview(worldRenderer, id, reference, owner);
 
                        Add(preview, initialSetup);
```
It turns out that SpawnMapActors already has a workaround:

https://github.com/OpenRA/OpenRA/blob/11e5d19f326edf9f26badde030e1e4aae31f0e07/OpenRA.Mods.Common/Traits/World/SpawnMapActors.cs#L38-L41

so we should make sure that both cases are equivalent (either spawn with world owner or not spawn at all).

**Comment by PunkPun** [comment]
So like this https://github.com/OpenRA/OpenRA/compare/bleed...PunkPun:OpenRA:fix-editor4

I also needed to replace the OwnerInit so that all other code to functions properly.

I could quickly make a PR if needed

**Comment by pchote** [comment]
> So like this [bleed...PunkPun:OpenRA:fix-editor4](https://github.com/OpenRA/OpenRA/compare/bleed...PunkPun:OpenRA:fix-editor4)

Yeah, looks reasonable. Lets go with that if @Mailaender is happy to add it here?

File: OpenRA.Mods.Common/Traits/World/EditorActorLayer.cs [hunk 1]
```diff
@@ -200,6 +207,10 @@ void SyncMultiplayerCount()
 				{
 					Players.Players.Remove(name);
 					OnPlayerRemoved();
+
+					foreach (var preview in previews)
+						if (preview.Owner == kv.Value)
+							preview.Owner = worldOwner;
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/World/EditorActorLayer.cs]: I still disagree with the addition of this line. I think we should avoid doing this until a move tool has been implemented

**Review by PunkPun — APPROVED** [review]

File: OpenRA.Mods.Common/Traits/World/SpawnMapActors.cs [hunk 2]
```diff
@@ -35,10 +35,13 @@ public void WorldLoaded(World world, WorldRenderer wr)
 			{
 				var actorReference = new ActorReference(kv.Value.Value, kv.Value.ToDictionary());
 
-				// If there is no real player associated, don't spawn it.
-				var ownerName = actorReference.Get<OwnerInit>().InternalName;
-				if (!world.Players.Any(p => p.InternalName == ownerName))
-					continue;
+				// If an actor's doesn't have a valid owner transfer ownership to neutral
+				var ownerInit = actorReference.Get<OwnerInit>();
+				if (!world.Players.Any(p => p.InternalName == ownerInit.InstanceName))
```
Comment (AspectInteractive2) [inline OpenRA.Mods.Common/Traits/World/SpawnMapActors.cs:40]: Shouldn't you check whether `ownerInit` is null first before trying to check its member?
`if (ownerInit != null && !world.Players.Any(p => p.InternalName == ownerInit.InstanceName))`

Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/World/SpawnMapActors.cs:40]: no, all actors must have an owner field

Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/World/SpawnMapActors.cs:40]: though I guess this new behaviour would allow not having it

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/World/SpawnMapActors.cs:40]: `.Get` will throw if it doesn't exist.

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/84713d9c3c459e80824c2fd33f2c667b8eb88ed7...4fcf775816eaac914ce4cf70a6e3f4c969c4c013)

https://github.com/OpenRA/OpenRA/commit/d96ec21b958b02d4065a83ce728a11fc9c2f6ffe
https://github.com/OpenRA/OpenRA/commit/daf10c34a6d090ce54674ec01e2ca1f5d5f31164

**Pull request merged**